### ISSUE TITLE
Not working with Unity

### ISSUE THREAD

**Issue by smakarov** [CoderLine/alphaTab#360]
# Your environment

* Version used: .net master AlphaTab.nupkg
* Platform used: C#
* Rendering engine used: SVG
* Unity: 2019.2
* Operating System and version (desktop or mobile): WIndows 10

# Expected Results

i'am trying to use this code inside Unity 2019.2
`Score score = ScoreLoader.LoadScoreFromBytes(File.ReadAllBytes(filePath), new AlphaTab.Settings());

            var settings = new Settings();
            settings.Core.Engine = "skia";
            var renderer = new ScoreRenderer(settings);
          
            var images = new List<SKImage>();
            var totalWidth = 0;
            var totalHeight = 0;
            renderer.PartialRenderFinished += r =>
                                              {
                                                  images.Add((SKImage) r.RenderResult);
                                              };
            renderer.RenderFinished += r =>
                                       {
                                           totalWidth = (int) r.TotalWidth;
                                           totalHeight = (int) r.TotalHeight;
                                       };
            renderer.Render();
            var path = $"{Application.persistentDataPath}/test.png";
            if (Debug.isDebugBuild)
            {
                Debug.Log($"totalWidth = {totalWidth}");
                Debug.Log($"totalHeight = {totalHeight}");
            }`
I've also tried renderer.RenderScore(score, new []{0,1});

# Observed Results

totalWidth and totalHeight is always 0.

**Comment by Danielku15** [comment]
Please follow the issue templates. I'm not addressing issues anymore unless they follow the issue template and provide the necessary information in a structured way to work on them.

**Comment by smakarov** [comment]
> Please follow the issue templates. I'm not addressing issues anymore unless they follow the issue template and provide the necessary information in a structured way to work on them.

Sorry. I have edited the post.

**Comment by Danielku15** [comment]
Thanks a lot for updating. 

**tl;dr:** set the `renderer.Width` to the desired rendering width. 

**Long version:**
To me the issue seems related to the absence of any width input. On platforms like the web, WPF and WinForms the UI controls get their width from a so called "UI Facade" in alphaTab which link the rendering pipeline with the visuals on the screen. 

If you are using the low-level rendering API of alphaTab from Unity, you will need to tell alphaTab how large the available surface for rendering is on your own by setting the `ScoreRenderer.Width` property. 

At least this is valid for the [`page`](https://docs.alphatab.net/develop/reference/property/display-layoutmode) layout which fits the bars into the available width. 

If you use the `horizontal` layout, width and height are set automatically as the width is calculated based on the bars, and the height based on the staves. 

The width property used to be part of the settings but was deprecated in favor of the `UI Facade` approach unifying the user experience across platforms. The low level API (Score Renderer) lacks a bit of available documentation or user guide so it is easy to miss. 

Hope this helps setting up alphaTab in Unity.

**Comment by rhclayto** [comment]
@Danielku15 So do you think it is possible to use the managed Ui Facade for Unity, rather than the low-level API?

**Comment by Danielku15** [comment]
> So do you think it is possible to use the managed UI Facade for Unity, rather than the low-level API?

The `ManagedUiFacade` is just a base class unifying some basic things. I only have some outdated experience with Unity and lack a bit of insight on what UI component systems it offers. Unity UI looks interesting and it might work out to make a facade for it: https://docs.unity3d.com/Packages/com.unity.ugui@1.0/manual/index.html

The UI facade is mainly implemented by providing "bridge components" that map UI controls and events to an alphaTab version. As long as the UI framework provides the possibility to create dynamically UI elements for displaying images and the cursors a UI facade should be possible. 

I just fear that the usecases in Unity might be too specific to provide an all-in-one control like for the other platforms. So using the low level API to render PNGs for further usage should be fine.

**Comment by smakarov** [comment]
@Danielku15 Thanks, it renders png fine now.
Similar problem with this method 
`private string testTab = "\\title \"Canon Rock\"\r\n\\subtitle \"JerryC\"\r\n\\tempo 90\r\n.\r\n:2 19.2{v f} 17.2{v f} |\r\n15.2{v f} 14.2{v f}|\r\n12.2{v f} 10.2{v f}|\r\n12.2{v f} 14.2{v f}.4 :8 15.2 17.2 |\r\n14.1.2 :8 17.2 15.1 14.1{h} 17.2 |\r\n15.2{v d}.4 :16 17.2{h} 15.2 :8 14.2 14.1 17.1{b(0 4 4 0)}.4 |\r\n15.1.8 :16 14.1{tu 3} 15.1{tu 3} 14.1{tu 3} :8 17.2 15.1 14.1 :16 12.1{tu 3} 14.1{tu 3} 12.1{tu 3} :8 15.2 14.2 |\r\n12.2 14.3 12.3 15.2 :32 14.2{h} 15.2{h} 14.2{h} 15.2{h}14.2{h} 15.2{h}14.2{h} 15.2{h}14.2{h} 15.2{h}14.2{h} 15.2{h}14.2{h} 15.2{h}14.2{h} 15.2{h}";
Score score = ScoreLoader.LoadScoreFromBytes(Encoding.ASCII.GetBytes(testTab));`
I got error on LoadScoreFromBytes method.

> NullReferenceException: Object reference not set to an instance of an object
AlphaTab.Importer.Gp3To5Importer.ReadScore () (at <3371ad0add004371adefaa18a99a453b>:0)
AlphaTab.Importer.ScoreLoader.LoadScoreFromBytes (System.Byte[] data, AlphaTab.Settings settings) (at <3371ad0add004371adefaa18a99a453b>:0)
AlphaTab.Test.Test+<StartAsync>d__7.MoveNext () (at Assets/Scripts/Test.cs:59)
--- End of stack trace from previous location where exception was thrown ---
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () (at <567df3e0919241ba98db88bec4c6696f>:0)
System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__6_0 (System.Object state) (at <567df3e0919241ba98db88bec4c6696f>:0)
UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at C:/buildslave/unity/build/Runtime/Export/Scripting/UnitySynchronizationContext.cs:150)
UnityEngine.UnitySynchronizationContext:ExecuteTasks() (at C:/buildslave/unity/build/Runtime/Export/Scripting/UnitySynchronizationContext.cs:104)

**Comment by Danielku15** [comment]
Same topic as [here](https://github.com/CoderLine/alphaTab/issues/351), pass in a settings object as second parameter and the exception disappears. .

**Comment by rhclayto** [comment]
> The UI facade is mainly implemented by providing "bridge components" that map UI controls and events to an alphaTab version. As long as the UI framework provides the possibility to create dynamically UI elements for displaying images and the cursors a UI facade should be possible.
> 
It does look like the Unity UI is scriptable in those ways. Supposing I wanted to try creating a UI facade to link AlphaTab to Unity UI, & use one of the current UI facades as a starting point or guide, which of the currently existing UI facades do you think is closest? It looks like there are ones for WinForms, WPF, Xamarin Android, & Xamarin Forms. Maybe you have a guess; if not, no worries.

By the way, very cool that you added Unity as a milestone! Thanks for your great feedback. Cool project!

**Comment by Danielku15** [comment]
> Supposing I wanted to try creating a UI facade to link AlphaTab to Unity UI, & use one of the current UI facades as a starting point or guide, which of the currently existing UI facades do you think is closest? 

This might depend on the way Unity UI works. The WPF and WinForms facades are both quite similar. Likely the WinForms is a better fit. It has a bit lesser features than WPF but is also more flexible as it does not rely on too much UI features being available.  

> By the way, very cool that you added Unity as a milestone! Thanks for your great feedback. Cool project!

This might be a small misunderstanding. I added this particular issue as to the milestone as this open topic should be closed until then. I did not consider a full unity support part as part of the release. 😉

**Comment by rhclayto** [comment]
> This might be a small misunderstanding. I added this particular issue as to the milestone as this open topic should be closed until then. I did not consider a full unity support part as part of the release. 😉

Gotcha. Thanks for the advice!

**Comment by Danielku15** [comment]
I'm closing this issue as there are no open discussion points in this issue. I entered https://github.com/CoderLine/alphaTabWebsite/issues/11 to add some better docs on using the low level API to eliminate confusion in future. If there are other dedicated issues regarding Unity popping up, feel free to report them with a new issue.
